### DISCUSSION TITLE
Short generated class names

### DISCUSSION THREAD

**Discussion by kireerik** [emotion-js/emotion#2172] [General]
Is there an option to generate short class names? (As short as possible. For example: a, b, c, . . ., aa, ab, ac . . .)

**Comment by Andarist** [comment]
No, this is not an option. Generated class names are based on a hash of the content of a ruleset

**Reply by kireerik** [reply]
I see, thank you!

Would this be possible or using a hash is necessary for some reason?

**Reply by Andarist** [reply]
Its required in our architecture - its a way to know if things have already been inserted into the CSSOM, allows for automatic rehydration when using SSR and more

**Reply by kireerik** [reply]
So if I understand correctly, this is only required on the client side (if Emotion is used there as well).

I am planning to use Emotion on the server side only (or rather simply at build time generating static pages). So in this case it could make sense and there is possibly no other reasons for using hashes instead of an incremental index.

Maybe a custom serializer is needed similar to https://emotion.sh/docs/testing.

**Reply by Andarist** [reply]
If u only plan to use SSR without any client side code handling styles then it should be possible to process rendered output and replace generated class names for shorter ones.

**Reply by kireerik** [reply]
Yes, definitely.

I think it would be nice to have an option where Emotion already outputs shorter class names.

**Reply by idoshamun** [reply]
But is it possible to at least remove the name of the component, like in the following example:
`css-qglrh0-FeedPage`
I'm thinking about something like `displayName` setting of styled-components:
https://styled-components.com/docs/tooling#better-debugging

**Reply by Andarist** [reply]
By default, labels should only be appended to the generated class names in dev mode (I'm talking about Emotion 11).

**Reply by idoshamun** [reply]
But for some reason, they appended in production as well. I use emotion with nextjs and preact.
I tried to change the label configuration in the babel plugin but it didn't change anything...
I mostly use the styled way, for what it's worth.

**Reply by Andarist** [reply]
Please always try to share a repro case in a runnable form - either by providing a git repository to clone or a codesandbox. OSS maintainers usually can't afford the time to set up the repro, even if exact steps are given.

**Reply by idoshamun** [reply]
Sorry for that. I'll do my best to create a repro soon.

**Reply by idoshamun** [reply]
I created a repro case on GitHub and deployed it on Vercel.
Here's the source:
https://github.com/idoshamun/emotion-labels-repro
And here's the deployed version:
https://emotion-labels-repro.vercel.app/

I created a nextjs using `create-next-app`, I added the emotion babel plugin, and created a div element using the styled function. To see the issue inspect the div element and see that the class name includes `StyledDiv` which is the label.
Let me know if you need more information.

**Reply by Andarist** [reply]
It seems that probably this PR fixes the problem: https://github.com/emotion-js/emotion/pull/2157 . I'll try to revive this in a moment and merge this in.

**Reply by idoshamun** [reply]
Awesome, thanks! Looking forward to the next release :)

**Reply by charrondev** [reply]
Out of curiosity have you compared and found a performance difference between shorter vs longer class names?

I feel like it would end up being pretty minimal if you are using `gzip` or `brotli` compression when serving the assets but I could be wrong.

**Reply by idoshamun** [reply]
Sorry for the late reply, I haven't done any performance comparison.

Actually, I updated the dependencies in my repro but it seems that the issue persists.
@Andarist could you please check it?

**Reply by Andarist** [reply]
@idoshamun your repro has not been updated - at least not on the remote. I've updated it myself though and it seems to work now.

Production (`npm run build && npm start`):
<img width="485" alt="Screenshot 2021-02-27 at 21 41 01" src="https://user-images.githubusercontent.com/9800850/109399723-d5678380-7944-11eb-9230-31fcc37b5331.png">
Development (`npm run dev`):
<img width="558" alt="Screenshot 2021-02-27 at 21 41 53" src="https://user-images.githubusercontent.com/9800850/109399739-eca67100-7944-11eb-8955-6d368b4c5fef.png">

**Reply by idoshamun** [reply]
You are right. Sorry for the trouble. It seems to work :)
Thanks for everything.

**Reply by kireerik** [reply]
> By default, labels should only be appended to the generated class names in dev mode (I'm talking about Emotion 11).

For me the labels are not included in `development` mode (either). However that would be nice to have (in `development` mode and or when and if I enable it).

This is my setup basically:
- https://github.com/emotion-js/emotion/issues/2118#issuecomment-1572296147

**Comment by kireerik** [comment]
I have implemented short class names. Here is a full example:
- # [refo](https://github.com/kireerik/refo)
&nbsp;
example:
- [getShortName.js](https://github.com/kireerik/refo/blob/48072d16f27d007e5c16d628cec1c00a91a67235/example/index/getShortName.js)

- [style/css.js#L22](https://github.com/kireerik/refo/blob/48072d16f27d007e5c16d628cec1c00a91a67235/example/index/style/css.js#L22)
